### PR TITLE
Fix api gateway path matching

### PR DESCRIPTION
**Pull request by Ankcorn**
This is a copy of [trpc/pull/2082](https://github.com/trpc/trpc/pull/2082) 

Big thanks to @mmkal and @Lilja for guiding me on this in https://github.com/trpc/trpc/pull/2050

Currently trpc only works when configured as the $default route, this makes it harder to integrate trpc into an existing aws project without the workaround mentioned by @adamsullovey in this [comment](https://github.com/trpc/trpc/pull/2050#issuecomment-1172907254) 

I did not alter the api gateway v1 behaviour, I don't use it so I am not really motivated to work on it.

What the codechange does: 

First I find all the matches in the routeKey against this regex `\{(.*?)\}`. This looks for anything between { }.

If no match is found an empty iterable is returned so it skips to the old logic.

the routeKey can look like

* $default
* any /trpc/{path+}
* any /trpc/{abc}/{path+}
* any /trpc/{path}

for each match we are looking for the greedy route (the one with the + in)
If we find the greedy route that will be the path for the trpc query so we return it. If the path is undefined we return an empty string. This will make trpc throw a 404 and the error message saying no query procedure in path ''.

If there is no greedy route we use the existing logic because we are assuming the user wants to match against the whole path in their project.

### CODE REVIEW

**Comment by adamsullovey** [comment]
Thank-you for following through with this! I'll give this branch a try while it is under review.

**Comment by adamsullovey** [comment]
It looks like this will work for URLs like https://somerandomcharacters.cloudfront.net/api/trpc/some-query which matches my use-case, so I'm excited.

**Comment by KATT** [comment]
cc @Lilja

**Comment by Ankcorn** [comment]
I'm not using trpc in production yet so might be slightly premature 😅

Maybe once this has been reviewed and merged you could reconsider. 

I'm in no rush for this to be merged so don't worry about the delay while @Lilja @mmkal find time to review

**Comment by Lilja** [comment]
What happens if you have `routeKey: "$default"`(serverless conf: `- httpApi: *`) but in your routes you would use `query("whatever")`. Then in your client you set it to `<url>/trpc` `rawPath` would be `"/trpc/whatever"` and it wouldn't find the route if I'm reading the code right.

Maybe that isn't a big issue?

**Comment by Ankcorn** [comment]
Yeah that would error. I think that's the desired outcome for that scenario.

**Comment by Lilja** [comment]
I did think that this solution might not be the best one, and I was thinking how to do it through complicated regexes. However, after sleeping on it, I think I'm quite fond of it.

We should add a test case with a query with a slash in it to really to get a use case like how @mmkal's [team uses tRPC ](https://github.com/trpc/trpc/pull/2050#discussion_r908733848)with slashes as some sort of nesting?

API Gateway documentation states that there cannot be more than one greedy wildcard in the routing to the lambda. So us finding `{whateverThisCouldBe+}` and accessing it through the `pathParameters` should be a valid solution.

I think this solution can also be deployed to the v1 rest api, just to replace a few parameters since they are not named exactly the same. I can help out to test this.

**Comment by Ankcorn** [comment]
That test is here :) https://github.com/trpc/trpc/blob/ca8a57c537b80eadbb46753630a55ec0bd83fc14/packages/server/test/adapters/awsLambda.test.tsx#L192

The reason we need to fallback to the initial implementation if no params are found in the routekey is that pathParameters is undefined.

You are right in theory this should work for rest api's I had not spotted that, I will update this PR this afternoon

**Review by skovhus — APPROVED** [review]
This looks good to me. ☀️

Not related to this PR, but relevant for the comments here: we are also using this adapter for Function URLs (so without an API Gateway). It would be great to update the documentation and the code to make it clear that this works for Function URLs.

**Comment by Lilja** [comment]
> That test is here :)

That test in particular uses `$routeKey: "default"`, which is will not trigger the greedy-plus-sign-regex and just slice the first character off. I would like to see a test where `routeKey` is something like `ANY /trpc/{param+}` and then the `rawPath` to be `/trpc/hello/darkness/my/old/friend`. That would test whether or `pathParams` can be used 😃

**Comment by Ankcorn** [comment]
@Lilja ah that's fair enough, happy to write this.

**Comment by Ankcorn** [comment]
Created a little demo using this branch

https://dam4bnpo075ad.cloudfront.net/articles
Source here https://github.com/Ankcorn/ssttrpc

Interestingly I ran into this issue https://github.com/trpc/trpc/discussions/2095 when doing so but editing the api gateway config to work around it was easy

**Comment by Ankcorn** [comment]
Hey it would be great to either review and merge as is or get some help with the rest api and function api usecases. I don't use them so deploying to get accurate event samples is a total chore.api and function api usecases. I don't use them so deploying to get accurate event samples is a total chore.


If anyone wants to do that for me and paste them in here I would be very thankful and happy to finish this off

**Comment by Lilja** [comment]
Here you go:

<details>

<summary>Click to expand</summary>

```
{
  body: null,
  headers: {
    'content-type': 'application/json',
    Accept: '*/*',
    'User-Agent': 'node-fetch/1.0 (+https://github.com/bitinn/node-fetch)',
    'Accept-Encoding': 'gzip,deflate',
    Connection: 'close',
    Host: 'localhost:4050'
  },
  httpMethod: 'GET',
  isBase64Encoded: false,
  multiValueHeaders: {
    'content-type': [ 'application/json' ],
    Accept: [ '*/*' ],
    'User-Agent': [ 'node-fetch/1.0 (+https://github.com/bitinn/node-fetch)' ],
    'Accept-Encoding': [ 'gzip,deflate' ],
    Connection: [ 'close' ],
    Host: [ 'localhost:4050' ]
  },
  multiValueQueryStringParameters: { batch: [ '1' ], input: [ '{"0":{"name":"Erik"}}' ] },
  path: '/leading/prefix//greet',
  pathParameters: { proxy: '/greet' },
  queryStringParameters: { batch: '1', input: '{"0":{"name":"Erik"}}' },
  requestContext: {
    accountId: 'offlineContext_accountId',
    apiId: 'offlineContext_apiId',
    authorizer: {
      claims: undefined,
      scopes: undefined,
      principalId: 'offlineContext_authorizer_principalId'
    },
    domainName: 'offlineContext_domainName',
    domainPrefix: 'offlineContext_domainPrefix',
    extendedRequestId: 'cl5htiwtg0000h2tw6mh1cc7d',
    httpMethod: 'GET',
    identity: {
      accessKey: null,
      accountId: 'offlineContext_accountId',
      apiKey: 'offlineContext_apiKey',
      apiKeyId: 'offlineContext_apiKeyId',
      caller: 'offlineContext_caller',
      cognitoAuthenticationProvider: 'offlineContext_cognitoAuthenticationProvider',
      cognitoAuthenticationType: 'offlineContext_cognitoAuthenticationType',
      cognitoIdentityId: 'offlineContext_cognitoIdentityId',
      cognitoIdentityPoolId: 'offlineContext_cognitoIdentityPoolId',
      principalOrgId: null,
      sourceIp: '127.0.0.1',
      user: 'offlineContext_user',
      userAgent: 'node-fetch/1.0 (+https://github.com/bitinn/node-fetch)',
      userArn: 'offlineContext_userArn'
    },
    operationName: undefined,
    path: '/leading/prefix//greet',
    protocol: 'HTTP/1.1',
    requestId: 'cl5htiwtg0001h2tw3l8w3l98',
    requestTime: '12/Jul/2022:08:54:33 +0200',
    requestTimeEpoch: 1657608873887,
    resourceId: 'offlineContext_resourceId',
    resourcePath: '/dev/leading/prefix/{proxy*}',
    stage: 'dev'
  },
  resource: '/leading/prefix/{proxy*}',
  stageVariables: null
}

```
</details>

```TypeScript
//server.ts
.query("greet", { input: z.object({name: z.string()}), resolve(req) { return req.inp.name })

// client.ts
client("http://localhost:4050/dev/leading/prefix/").query("greet", { name: "Erik" })
```

```
# serverless.yml
rest-api-leading-prefix:
    # ...
    events:
        - http
            path: /leading/prefix/{proxy+}
            method: any 

```

**Comment by Lilja** [comment]
I think the double slash-thing can be ignored, opened an issue for it. #2158

**Comment by Ankcorn** [comment]
Thank you @Lilja. Have added the code to support that scenario here https://github.com/trpc/trpc/pull/2087/commits/22e4a64586bffbd9fb7823616e752bde91693555

Having the example request from serverless offline only makes me a little nervous, I have been burned a few times from issues where request bodies differ once deployed

File: packages/server/src/adapters/aws-lambda/index.ts
Comment (Ankcorn) [reply packages/server/src/adapters/aws-lambda/index.ts:59]: Interesting that it took offense to `group.replace('*', '')` and not `group.replace('+', '')` :laughing:

Comment (Lilja) [reply packages/server/src/adapters/aws-lambda/index.ts:59]: I wonder if you need to escape both? Doesn't look like it's harmful: 
image

**Comment by Lilja** [comment]
> Having the example request from serverless offline makes me a little nervous, I have been burned a few times from issues where request bodies differ once deployed

Absolutely agree, however. serverless offline is very battle tested and have gone through alot of major versions. I think it's safe to trust these inputs.

File: packages/server/src/adapters/aws-lambda/index.ts
Comment (Ankcorn) [reply packages/server/src/adapters/aws-lambda/index.ts:59]: Eslint doesn't like that. Because this is a trusted input (i.e. user cannot use this to execute an unwanted trpc query) I would suggest that this is a false positive and my code is fine

**Comment by Ankcorn** [comment]
> battle tested

I think my latest fix contains a bug actually :laughing: 
https://github.com/dherault/serverless-offline/issues/1431

**Comment by Lilja** [comment]
Fair enough, I spun up an actual REST API and you are right. A bit concerning...

<details>

<summary>Click to expand</summary>

```
{
  resource: '/leading/prefix/{proxy+}',
  path: '/leading/prefix/test/thing',
  httpMethod: 'GET',
  headers: {
    accept: '*/*',
    Host: 'censored',
    'User-Agent': 'curl/7.83.1',
    'X-Amzn-Trace-Id': 'Root=1-62cd6757-7259b8dd7b8f145033708acd',
    'X-Forwarded-For': 'censored',
    'X-Forwarded-Port': '443',
    'X-Forwarded-Proto': 'https'
  },
  multiValueHeaders: {
    accept: [ '*/*' ],
    Host: [ 'censored' ],
    'User-Agent': [ 'curl/7.83.1' ],
    'X-Amzn-Trace-Id': [ 'Root=1-62cd6757-7259b8dd7b8f145033708acd' ],
    'X-Forwarded-For': [ 'censored' ],
    'X-Forwarded-Port': [ '443' ],
    'X-Forwarded-Proto': [ 'https' ]
  },
  queryStringParameters: null,
  multiValueQueryStringParameters: null,
  pathParameters: { proxy: 'test/thing' },
  stageVariables: null,
  requestContext: {
    resourceId: 'censored',
    resourcePath: '/leading/prefix/{proxy+}',
    httpMethod: 'GET',
    extendedRequestId: 'VJ0VvFLiIAMFiOQ=',
    requestTime: '12/Jul/2022:12:21:43 +0000',
    path: '/dev/leading/prefix/test/thing',
    accountId: '328886342752',
    protocol: 'HTTP/1.1',
    stage: 'dev',
    domainPrefix: 'qtdldn2207',
    requestTimeEpoch: 1657628503765,
    requestId: 'ff7a0efa-97df-4a02-abee-cbaa1927c0a5',
    identity: {
      cognitoIdentityPoolId: null,
      accountId: null,
      cognitoIdentityId: null,
      caller: null,
      sourceIp: 'censored',
      principalOrgId: null,
      accessKey: null,
      cognitoAuthenticationType: null,
      cognitoAuthenticationProvider: null,
      userArn: null,
      userAgent: 'curl/7.83.1',
      user: null
    },
    domainName: 'censored',
    apiId: 'censored'
  },
  body: null,
  isBase64Encoded: false
}

```

</details>

`curl https://[...]/dev/leading/prefix/test/thing`

**Review by Lilja — APPROVED** [review]
Just don't merge without squashing.

**Comment by Ankcorn** [comment]
I have updated the serverless offline issue with the info we found here :)

**Comment by Ankcorn** [comment]
@KATT I believe after all my faffing about this PR is ready for you! 😁

**Review by skovhus — COMMENTED** [review]
Happy to see the attention to detail and thorough testing. 👏

File: packages/server/src/adapters/aws-lambda/index.ts
Comment (skovhus) [inline packages/server/src/adapters/aws-lambda/index.ts:66]: Wouldn’t it make sense to extract the shared code here into a function taking routeKey or resource depending on the payload version?

Comment (Ankcorn) [reply packages/server/src/adapters/aws-lambda/index.ts:66]: All the options to refactor here feel like they makes the code more complicated not less which is why its just duplicated right now, if you have any suggestions happy to take them on board :)

**Pull request merged**
